air/src/stack/op_flags/mod.rs

```diff
@@ -61,13 +79,275 @@ impl<E: FieldElement> OpFlags<E> {
+        // to the next frame.
+        let f0000 = degree7_op_flags[0];
+        // flag of prefix when the first 4 elements in op_bits are `0100`. Caching the result
+        // for the compution of the left_shift_2. It's a flag for degree 7 operation which has
+        // shifted the stack to the left from third element onwards.
+        let f0100 = degree7_op_flags[32];
+
+        // adding the fifth most significant bit.
+        for i in (0..64).step_by(8) {
+            degree7_op_flags[i + 4] = degree7_op_flags[i] * frame.op_bit(2);
+            degree7_op_flags[i] *= not_2;
+        }
+
+        // adding the sixth most significant bit.
+        for i in (0..64).step_by(4) {
+            degree7_op_flags[i + 2] = degree7_op_flags[i] * frame.op_bit(1);
+            degree7_op_flags[i] *= not_1;
+        }
+
+        // flags of all the mov{up/dn}{2-8}, swapw{2-3} operations.
+        let mov2_flag = degree7_op_flags[10];
+        let mov3_flag = degree7_op_flags[12];
+        let mov4_flag = degree7_op_flags[16];
+        let mov5_flag = degree7_op_flags[18];
+        let mov6_flag = degree7_op_flags[20];
+        let mov7_flag = degree7_op_flags[22];
+        let mov8_flag = degree7_op_flags[26];
+        let swapwx_flag = degree7_op_flags[28];
+
+        // adding the least significant bit.
+        for i in (0..64).step_by(2) {
+            degree7_op_flags[i + 1] = degree7_op_flags[i] * frame.op_bit(0);
+            degree7_op_flags[i] *= not_0;
+        }
+
+        // flag when the items from first point onwards are copied over. It doesn't have noop.
+        let no_change_1_flag = f0000 - degree7_op_flags[0];
+        // flag when the items from second point onwards are shifted to the left. It doesn't have assert.
+        let left_change_1_flag = f0100 - degree7_op_flags[32];
+
+        let helper = frame.op_bit_helper();
+
+        // computation of degree 6 and 4 flags.
+        degree4_op_flags[0] = not_3 * not_2;
+        degree4_op_flags[1] = frame.op_bit(2) * not_3;
+        degree4_op_flags[2] = not_2 * frame.op_bit(3);
+        degree4_op_flags[3] = frame.op_bit(3) * frame.op_bit(2);
+        degree4_op_flags.copy_within(0..4, 4);
+
+        for i in 0..4 {
+            degree4_op_flags[i] *= not_4;
+            degree6_op_flags[2 * i] = degree4_op_flags[i];
+        }
+        for i in 4..8 {
+            degree4_op_flags[i] *= frame.op_bit(4);
+            degree6_op_flags[2 * i] = degree4_op_flags[i];
+        }
+
+        // helper register is multiplied with the intermediate values to enumerate all the possible
+        // degree 4 operations flags.
+        degree4_op_flags
+            .iter_mut()
+            .take(8)
+            .for_each(|v| *v *= helper);
+
+        // flag of END operation. It's effect on stack depends if the current block being executed is
+        // a loop block or not.
+        let end_flag = degree4_op_flags[4] * frame.prev_loop();
```

Comment:
I think the naming `end_flag` is confusing. I think think what this flag actually means is `shift_left_on_end`, right?